### PR TITLE
Fixes #4360 by introducing a ContainerSpec which gets implemented by …

### LINKED ISSUE #4360: Secret Interface for pods

### PR DESCRIPTION
**Pull request by aquamatthias**
…the MesosContainer and AppDefinition.

Also change the plugin interface of RunSpecTaskProcessor.

### CODE REVIEW

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/RunSpec.scala [hunk 1]
```diff
@@ -1,14 +1,47 @@
 package mesosphere.marathon.plugin
 
 /**
-  * A Marathon RunSpec Definition
+  * A Marathon Container definition
   */
-trait RunSpec {
-  def id: PathId
+trait ContainerSpec {
```
Comment (jasongilanfarr) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/RunSpec.scala]: vals not defs.

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/RunSpec.scala [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
+  /**
+    * The labels in that container
+    */
   def labels: Map[String, String]
+}
+
+/**
+  * A Marathon RunSpec Definition
+  */
+trait RunSpec {
```
Comment (jasongilanfarr) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/RunSpec.scala]: vals, not defs.

File: src/main/scala/mesosphere/marathon/core/pod/MesosContainer.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    case state.EnvVarSecretRef(secret) => raml.EnvVarSecretRef(secret)
+  }
+
+  def apply(c: PodContainer): MesosContainer = MesosContainer(
+    name = c.name,
+    exec = c.exec,
+    resources = c.resources,
+    endpoints = c.endpoints,
+    image = c.image,
+    env = c.environment.fold(Map.empty[String, state.EnvVarValue])(_.values.map{ case (k, v) => k -> toStateEnv(v) }),
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/pod/MesosContainer.scala]: not using a helper like the others?

File: src/main/scala/mesosphere/marathon/state/RunSpec.scala [hunk 4]
```diff
@@ -75,4 +75,5 @@ trait RunSpec extends plugin.RunSpec {
   // TODO(
   val fetch = Seq.empty[FetchUri]
   val portNames = Seq.empty[String]
+  def user: Option[String]
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/state/RunSpec.scala]: val

**Comment by jasongilanfarr** [comment]
I don't see the value of this change. Each pod container may have different users and environment variables etc. While I think its great that we are trying to be as generic as possible, I think we're going way too far with it and actually making mistakes instead of pattern matching when they are _different_. I put the TODOs at the bottom of RunSpec trying to say "we should remove _all of these fields_" and pattern match when they are necessary.

**Comment by aquamatthias** [comment]
@jasongilanfarr 
1) vals versus def: http://stackoverflow.com/questions/19642053/when-to-use-val-or-def-in-scala-traits
The recommendation is to use def in the abstract trait, which can be implemented as vals (which is what we do)
2) state.RunSpec I totally agree. We need to get rid of this. The added method here has been deleted in the plugin.RunSpec (external interface) and added to the state.RunSpec (internal interface). 

The value of this change is: have a plugin interface, that allows for TaskInfo/TaskGroup customization that is needed for the EE Plugins.

**Comment by jasongilanfarr** [comment]
I've read that article before and while interesting the restriction of
stable value is indicative of the intent. Val says it will never change,
def says it can.

As for the plugin interface, the base trait in the plugin probably needs
something more so that pods and apps can be differentiated. I don't think
container is the right approach, since pods can have many and apps may not
even have one. I'm inclined to suggest having appspec and podspec traits in
the plugin interface.

**Comment by aquamatthias** [comment]
Split the RunSpec into ApplicationSpec and PodSpec.

File: src/main/scala/mesosphere/marathon/core/pod/MesosContainer.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  volumeMounts: scala.collection.immutable.Seq[VolumeMount] = Nil,
+  artifacts: scala.collection.immutable.Seq[Artifact] = Nil, //TODO(PODS): use FetchUri
+  labels: Map[String, String] = Map.empty,
+  lifecycle: Option[Lifecycle] = None) extends ContainerSpec
+
+object MesosContainer {
+
+  //TODO(PODS): find a better place for this converters (should probably live in the API)
+
+  def toStateEnv(envVars: EnvVars): Map[String, state.EnvVarValue] = {
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/pod/MesosContainer.scala]: We should probably have the conversion in state.EnvVarValue

**Comment by jasongilanfarr** [comment]
lgtm

File: plugin-interface/src/main/scala/mesosphere/marathon/plugin/task/RunSpecTaskProcessor.scala [hunk 6]
```diff
@@ -1,18 +1,25 @@
...
-import mesosphere.marathon.plugin.RunSpec
+import mesosphere.marathon.plugin.{ ApplicationSpec, PodSpec }
 import mesosphere.marathon.plugin.plugin.Plugin
-import org.apache.mesos.Protos.TaskInfo
+import org.apache.mesos.Protos.{ TaskGroupInfo, TaskInfo }
 
 /**
   * RunSpecTaskProcessor mutates a Mesos task info given some app specification.
   */
 trait RunSpecTaskProcessor extends Plugin {
```
Comment (jdef) [inline plugin-interface/src/main/scala/mesosphere/marathon/plugin/task/RunSpecTaskProcessor.scala]: nit: probably need a better name for this

**Comment by jdef** [comment]
nit, otherwise lgtm

**Comment by aquamatthias** [comment]
Rebased and merged

**Pull request merged**